docs/proposal/20220712-garbage-collection.md

```diff
@@ -0,0 +1,338 @@
+- Delete CAPA owned AWS resources for that cluster that are related to the **NetworkSpec**. This will be done via the existing network service (11,12).
+
+###### New Garbage Collection Service
+
+For cluster deletion there will be a **ReconcileDelete** function. The purpose of this function is to identify the AWS resources that have been created for the cluster via the CCM. And then for the identified resources delete them in AWS.
+
+To identify the resources that the CCM has created for the cluster we will use the **AWS resource tagging api** to query for all  resources that have a label called `kubernetes.io/cluster/[CLUSTERNAME]` with a value of `owned`. Note `[CLUSTERNAME]` will be replaced with the Kubernetes cluster name.
+
+Based on the list of resources returned we will group these by the owing AWS service (i.e. **ec2**, **elasticloadbalancing**). The grouped resources will then be passed to a function for that service which will take care of cleaning up the resources in AWS via API calls (8).
+
+The reason we are grouping by AWS service is that order can matter when deleting. For example, with the **elasticloadbalancing** service you need to delete the load balancers before any target groups.
+
+We will need to create the gc service so that its easy to add new cleanup functions for services in the future [NFR8](#NFR8).
+
+Once all the resources have been cleaned up we can remove the `awsexternalresourcegc.infrastructure.cluster.x-k8s.io` finalizer (9) to indicate that the cluster has been garbage collected.
+
+> NOTE: we will initally not handle clean-up of EBS volumes due to the potential risk of accidental data deletion. This will be considered for a future enhancement.
+
+##### clusterawsadm changes
+
+We would like to supply a way for the user to manually mark a cluster as requiring garbage collection and vice versa opting out of garbage collection [FR6](#FR6).
+
+We will add 2 new commands to `clusterawsadm` to perform this:
+
+- **clusterawsadm gc enable** - this will add the new finalizer and annotation to the infra cluster object. 
+- **clusterawsadm gc disable** - this will remove the new finalizer and add/set the `aws.cluster.x-k8s.io/external-resource-gc` annotation to `false`.
+
+### Alternative Approaches Considered
+
+#### Using CCM to do the delete
+
+The initial implementation of the garbage collector relied on the CCM in the child/tenant cluster doing the delete. When a cluster is deleted CAPA would pause the delete reconciliation until garbage collection had been done.
+
+The garbage collector (a separate controller) would:
+
+- Connect to the tenant cluster and get a list of `Services` of type `LoadBalancer`.
+- Delete each of the `Services` of type `LoadBalancer`. The CCM in the child/tenant cluster at this point will delete the resources it created in AWS.
+- Requeue until all the services as deleted.
+- Once all the services are delete mark the cluster has having been garbage collected.
```

Comment:
That's a bit odd sentence, because "marking" here means that we remove the finaliser right? Or I have to read this again. Is there an annotation that's added saying: GC: done?

Reply:
Reworded, hopefully that is better.